### PR TITLE
[WIP] Dataset indexer

### PR DESCRIPTION
**Pull request by corochann**
This is work in progress to support `features` indexer for chainer dataset class.

Refer: 
[[New feature] Feature extraction support for Chainer dataset class](https://github.com/chainer/chainer/issues/3251)

### CODE REVIEW

File: chainer/dataset/dataset_mixin.py [hunk 1]
```diff
@@ -84,3 +88,68 @@ def get_example(self, i):
...
+        This method may be override to support efficient feature extraction.
+
+        Args:
+            i (int): `i`-th data to be extracted
+            j (int): `j`-th feature to be extracted
+
+        Returns: feature
+
+        """
+        return self.get_example(i)[j]
```
Comment (delta2323) [inline chainer/dataset/dataset_mixin.py]: Can we assume that each sample of the dataset can accept indexing via integers?

File: chainer/datasets/dict_dataset.py [hunk 2]
```diff
@@ -39,3 +43,50 @@ def __getitem__(self, index):
...
+
+
+class DictDatasetFeaturesIndexer(BaseFeaturesIndexer):
+    """FeaturesIndexer for TupleDataset"""
+
+    def __init__(self, dataset):
+        """
+
+        Args:
+            dataset (TupleDataset): TupleDataset instance
```
Comment (delta2323) [inline chainer/datasets/dict_dataset.py]: DictDataset?

**Comment by delta2323** [comment]
Are you thinking of supporting indexers for `TupleDataset` and `DictDataset`?

**Comment by corochann** [comment]
I already implemented the support of `dict_dataset` as well.
https://github.com/chainer/chainer/pull/3252/files#diff-afabfc7ee7723839c8042dc74a3b23b5

As you can see, I'm going to implement for `DatasetMixin`, `ImageDataset`, `LabeledImageDataset` class too.

**Comment by corochann** [comment]
Remaining task is to 
- For DatasetMixin class, I think it is better for default behavior to create a cache to return features.
  (currently, it is calling `get_example` every time in extract_feature)
- write a test

Once I got positive comment for the idea, I will proceed above implementation.

**Comment by delta2323** [comment]
It is a handy feature as I frequently do like `labels = [d[1] for d in dataset]`. So I want to proceed the PR. By the way, is it possible to separate it into smaller ones so that we can review it smoothly?

**Comment by corochann** [comment]
I separated the PR, and at first submit a PR for only TupleDataset.

#3343

**Comment by Hakuyume** [comment]
I'm implementing `PickableDataset` (https://github.com/chainer/chainercv/pull/454) which is another solution for #3251. You can see the usage of it [here](https://github.com/chainer/chainercv/pull/454#issuecomment-336684256). I would like to migrate this dataset from ChainerCV to Chainer because it is not only for CV. With this dataset, we can implement efficient feature selection more easily than Indexer.

**Comment by corochann** [comment]
@Hakuyume 
Thank you for your comment, I don't know similar discussion is going on in chainer cv (while I was making PR from almost 2 month ago...).

1. One important use case for me is that I want to extract label information after splitting the dataset into train and validation, which is `SubDataset` (subclass of `DatasetMixin`), as written in this discussion,
Therefore, even specific dataset class support this feature, I cannot extract labels information efficiently when I split the dataset.
https://github.com/chainer/chainer/pull/3343

2. I guess actually `DictDataset` can do the similar job, which `PickableDataset` want to support.
Please refer this test for demo.
https://github.com/chainer/chainer/pull/3252/files#diff-afabfc7ee7723839c8042dc74a3b23b5R48

3. I think you already noticed it, as you are discussing in the chainercv thread (https://github.com/chainer/chainercv/pull/454). 
One another motivation for this PR is to support efficient data extraction, which means, do not copy numpy array but use view reference when possible.
It can be done by overriding `extract_feature_by_slice` method.
https://github.com/chainer/chainer/pull/3343/files#diff-81b6edfb9c10c745e45d1990196ec9f6R73

One troublesome stuff is that if the user want to optimize their own DatasetMixin class, user need to re-implement their own `extract_feature` method or `extract_feature_by_slice` method.

**Comment by Hakuyume** [comment]
> 1. One important use case for me is that I want to extract label information after splitting the dataset into train and validation, which is SubDataset (subclass of DatasetMixin), as written in this discussion (#3343),
Therefore, even specific dataset class support this feature, I cannot extract labels information efficiently when I split the dataset.

As you pointed out, `SubDataset` erases pickable feature. 
One solution is adding `split` or `sub` method to `PickableDataset` which can make sub dataset with pickable feature. Another solution is modifying `SubDataset` to treat `PickableDataset` in different way.

> 2. I guess actually DictDataset can do the similar job, which PickableDataset want to support.
Please refer this test for demo.

`DictDataset` seems not to provide the feature I want. I want to support efficient access for multiple data.

Assume that `dataset` has three data types, `data_0`, `data_1` and `data_2`. The i-th values of `data_0` and `data_1` can be retrieved from a getter `get_0_1(i)`. And the i-th value of `data_2` can be retrieved from another getter `get_2(i)`.

In the case of `PickableDataset`.
```
sub_dataset = dataset.pick('data_0', 'data_1')
sub_dataset[i]  # get_0_1[0] is called only once
```

In the case of `DictDataset, I guess it will be as follows.
```
data_0, data_1 = dataset[:, ['data_0', 'data_1']]
sub_dataset = DictDataset(data_0=data_0, data_1=data_1)
sub_dataset[i]  # get_0_1[0] is called twice
```

> 3. I think you already noticed it, as you are discussing in the chainercv thread (chainer/chainercv#454).
One another motivation for this PR is to support efficient data extraction, which means, do not copy numpy array but use view reference when possible.
It can be done by overriding extract_feature_by_slice method.
https://github.com/chainer/chainer/pull/3343/files#diff-81b6edfb9c10c745e45d1990196ec9f6R73

Yes. That is a disadvantage of `PickableDataset`.

> One troublesome stuff is that if the user want to optimize their own DatasetMixin class, user need to re-implement their own extract_feature method or extract_feature_by_slice method.

I think that is a huge burden for users. With `PickableDataset`, it becomes more easy.

**Comment by corochann** [comment]
1. Actually, it is not only the case for `SubDataset`, but also when we used `ConcatenatedDataset` and `TransformDataset` etc. So I thought it was easier/user-friendly to introduce unified interface for all chainer dataset, rather than develop specific dataset class. 

2. Sorry, I could not get your point.
If you want to extract `i-th` 'data_0' and 'data_1' from DictDataset
```
data_0, data_1 = dict_dataset.features[i, ['data_0', 'data_1']]
```
is enough.
Also if you want to extract 100-200-th data for example, you can call
```
data_0, data_1 = dict_dataset.features[100:200, ['data_0', 'data_1']]
```
and it returns "view" (not copy) of the dataset.

3. Yeah I agree implementing one extra method for dataset_mixin class is troublesome for users.
But even if you don't override the method, the code works. (but it is just not optimized to return view, but copy can be returned.)
https://github.com/chainer/chainer/pull/3252/files#diff-6ec23c4ee85004cde0ad58ad3fa9877dR117

**Comment by Hakuyume** [comment]
> 2.  Sorry, I could not get your point.
  If you want to extract i-th 'data_0' and 'data_1' from DictDataset
  data_0, data_1 = dict_dataset.features[i, ['data_0', 'data_1']]
   is enough.
  Also if you want to extract 100-200-th data for example, you can call
  data_0, data_1 = dict_dataset.features[100:200, ['data_0', 'data_1']]

I want to get a dataset which contains both `data_0` and `data_1`. Not two datasets.

**Comment by corochann** [comment]
Ah I see.

I don't mind the backend method of how to achieve this feature, and I'm happy if you can proceed this feature support.
I just would like to request that it would be great if this feature can be applicable to extended dataset like `SubDataset`, `ConcatenatedDataset` and `TransformDataset` etc.

**Comment by mitmul** [comment]
@beam2d @Hakuyume @corochann we will consider a unified way to satisfy some related demands (https://github.com/chainer/chainer/pull/3976, https://github.com/chainer/chainer/pull/3343). Let us make this PR pending until we make the decision. (I changed the assignee to @delta2323 because he has already reviewed some part of this PR.)

**Comment by corochann** [comment]
I understand, please make it pending or closed.

**Comment by corochann** [comment]
Seems that Chainer team will consider another way to provide this functionality.

**Pull request closed without merging**